src/passes/InstrumentMemory.cpp

```diff
@@ -0,0 +1,182 @@
+  }, {
+    { i64_load8_u,  i64_load8_s   },
+    { i64_load16_u, i64_load16_s  },
+    { i64_load32_u, i64_load32_s  },
+    { i64_load,     i64_load      }
+  }, {
+    { empty                       },
+    { empty                       },
+    { f32_load,     f32_load      }
+  }, {
+    { empty                       },
+    { empty                       },
+    { empty                       },
+    { f64_load,     f64_load      }
+  }
+};
+
+Name i32_store8("i32_store8");
+Name i32_store16("i32_store16");
+Name i32_store("i32_store");
+Name i64_store8("i64_store8");
+Name i64_store16("i64_store16");
+Name i64_store32("i64_store32");
+Name i64_store("i64_store");
+Name f32_store("f32_store");
+Name f64_store("f64_store");
+
+Name stores [4][4] = {
+  { i32_store8,   i32_store16, i32_store              },
+  { i64_store8,   i64_store16, i64_store32, i64_store },
+  { empty,        empty,       f32_store              },
+  { empty,        empty,       empty,       f64_store }
+};
+
+static int32_t bytesToShift(unsigned bytes) {
+  switch (bytes) {
+    case 1: return 0;
+    case 2: return 1;
+    case 4: return 2;
+    case 8: return 3;
+    default: {}
+  }
+  abort();
+  return -1; // avoid warning
+}
+
+struct InstrumentMemory : public WalkerPass<PostWalker<InstrumentMemory>> {
+  void visitLoad(Load* curr) {
+    replaceCurrent(makeLoadCall(curr));
+  }
+  void visitStore(Store* curr) {
+    replaceCurrent(makeStoreCall(curr));
+  }
+  void addImport(Module *curr, Name name, std::string sig) {
+    auto import = new Import;
+    import->name = name;
+    import->module = ENV;
+    import->base = name;
+    import->functionType = ensureFunctionType(sig, curr)->name;
+    import->kind = ExternalKind::Function;
+    curr->addImport(import);
+  }
+  void visitModule(Module *curr) {
+    // Loads
+    addImport(curr, i32_load8_s,  "iiiii");
+    addImport(curr, i32_load8_u,  "iiiii");
+    addImport(curr, i32_load16_s, "iiiii");
+    addImport(curr, i32_load16_u, "iiiii");
+    addImport(curr, i32_load,     "iiiii");
+    addImport(curr, i64_load8_s,  "jiiii");
+    addImport(curr, i64_load8_u,  "jiiii");
+    addImport(curr, i64_load16_s, "jiiii");
+    addImport(curr, i64_load16_u, "jiiii");
+    addImport(curr, i64_load32_s, "jiiii");
+    addImport(curr, i64_load32_u, "jiiii");
+    addImport(curr, i64_load,     "jiiii");
+    addImport(curr, f32_load,     "fiiii");
+    addImport(curr, f64_load,     "diiii");
+
+    // Stores
+    addImport(curr, i32_store8,  "viiiii");
+    addImport(curr, i32_store16, "viiiii");
+    addImport(curr, i32_store,   "viiiii");
+    addImport(curr, i64_store8,  "viiiij");
+    addImport(curr, i64_store16, "viiiij");
+    addImport(curr, i64_store32, "viiiij");
+    addImport(curr, i64_store,   "viiiij");
+    addImport(curr, f32_store,   "viiiif");
+    addImport(curr, f64_store,   "viiiid");
+  }
+
+private:
+  Expression* makeLoadCall(Load* curr) {
+    static Index loadId = 0;
```

Comment:
Seems like this should probably just be a class-scoped variable rather than a static. I guess that having this ID also means that this pass should have `isFunctionParallel` be false, assuming we want unique IDs.